src/azul/plugins/anvil.py

```diff
@@ -0,0 +1,104 @@
+from abc import (
+    ABC,
+)
+from typing import (
+    AbstractSet,
+    Generic,
+    Iterable,
+    Optional,
+    TypeVar,
+    Union,
+)
+
+import attr
+from more_itertools import (
+    one,
+)
+
+from azul.indexer import (
+    BUNDLE_FQID,
+    Bundle,
+)
+from azul.indexer.document import (
+    EntityReference,
+    EntityType,
+)
+from azul.types import (
+    JSON,
+    MutableJSON,
+)
+
+# AnVIL snapshots do not use UUIDs for primary/foreign keys.
+# This type alias helps us distinguish these keys from the document UUIDs,
+# which are drawn from the `datarepo_row_id` column.
+# Note that entities from different tables may have the same key, so
+# `KeyReference` should be used when mixing keys from different entity types.
+Key = str
+
+
+@attr.s(frozen=True, auto_attribs=True, kw_only=True, slots=True)
+class KeyReference:
+    key: Key
+    entity_type: EntityType
+
+
+ENTITY_REF = TypeVar(name='ENTITY_REF', bound=Union[EntityReference, KeyReference])
```

Comment:
the kwarg confuses PyCharm

```suggestion
ENTITY_REF = TypeVar('ENTITY_REF', bound=Union[EntityReference, KeyReference])
```